### PR TITLE
Use list of URLs for configuring hosts for all the upstream hosts

### LINKED ISSUE #83: Use URIs for configuring hosts for reaching upstream hosts
Hosts for LoginManager, ServiceIdentifier, IdentityManager and AccessManager should be configured with a format of `<http[s]>://<host>[:<port>]`

### PR DESCRIPTION
**Pull request by maheshkelkar**
Fixes #83

### CODE REVIEW

**Comment by maheshkelkar** [comment]
@PeterFCaswell Please take a look at this.

File: core/src/main/scala/com/lookout/borderpatrol/Binder.scala [hunk 1]
```diff
@@ -17,7 +20,7 @@ object Binder {
    */
   trait BinderContext[A] {
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: Comments on this trait?

File: core/src/main/scala/com/lookout/borderpatrol/Binder.scala [hunk 2]
```diff
@@ -39,12 +42,20 @@ object Binder {
...
-      )).apply(req.req)
+      this.synchronized(cache.getOrElse(req.name, {
+        // If its https, use TLS
+        val https = !req.hosts.filter(u => u.getProtocol == "https").isEmpty
+        val hostname = req.hosts.map(u => u.getHost()).mkString
+        // Find CSV of host & ports
+        val hostAndPorts = req.hosts.map(u => u.getAuthority()).mkString(",")
+        util.Combinators.tap(
+          if (https) Httpx.client.withTls(hostname).newService(hostAndPorts)
+          else Httpx.newService(hostAndPorts)
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: I know that this is considered an appropriate style in Scala, but I find this incredibly cramped.
We're not paying for newlines. :-)

File: example/src/main/scala/com/lookout/borderpatrol/example/Config.scala [hunk 3]
```diff
@@ -164,28 +166,86 @@ object Config {
...
+      throw new InvalidConfigError(s"hosts configuration for ${name} in ${field}: has differing protocols")
+    // Make sure hosts in Manager have either http or https protocol
+    if (!hosts.map(_.getProtocol()).mkString.matches("http[s]*"))
+      throw new InvalidConfigError(s"hosts configuration for ${name} in ${field}: has unsupported protocol")
+    // Make sure https hosts have a matching hostname
+    if (!hosts.filter(u => u.getProtocol == "https").isEmpty &&
+      hosts.map(u => u.getHost()).size != 1)
+      throw new InvalidConfigError(
+        s"hosts configuration for ${name} in ${field}: https urls have mismatching hostnames")
+  }
```
Comment (PeterFCaswell) [inline example/src/main/scala/com/lookout/borderpatrol/example/Config.scala]: I find the layout of this code masks how this gets parsed. It looks at first glance that 'hosts.map' is something that happens if the previous line's case is true. Maybe a different layout or some curly brackets?

File: example/src/main/scala/com/lookout/borderpatrol/example/Config.scala [hunk 4]
```diff
@@ -164,28 +166,86 @@ object Config {
...
+
+  /**
+   * Validate serviceIdentifier configuration
+   * @param field
+   * @param sids
+   */
+  def validateServiceIdentifierConfig(field: String, sids: Set[ServiceIdentifier]): Unit = {
+    // Find if ServiceIdentifiers have duplicate entries
+    if (sids.size > sids.map(sid => (sid.path, sid.subdomain)).size)
+    throw new DuplicateConfigError("path and subdomain", "serviceIdentifiers")
```
Comment (PeterFCaswell) [inline example/src/main/scala/com/lookout/borderpatrol/example/Config.scala]: Can you indent this a couple of spaces, please?

File: core/src/main/scala/com/lookout/borderpatrol/Binder.scala (hunk 2 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: done

File: core/src/main/scala/com/lookout/borderpatrol/Binder.scala (hunk 1 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: I have this comment there. Do you think we should add to that?

  /**
   * The trait for the context binding. It exposes common methods
   * to be made available from all the contexts.
   * @tparam A
   */

Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: Oh. I'm sorry. I misread the diff output. It's okay.

**Pull request merged**